File: Data Structures & Algorithms/eating-bananas/submission-5.js
class Solution {
    /**
     * @param {number[]} piles
     * @param {number} h
     * @return {number}
     */
    minEatingSpeed(piles, h) {
        let l = 1;
        let r = Math.max(...piles)
        let result = r

        while (l <= r) {
            const k = Math.round((r + l) / 2)
            let total = 0

            for (const pile of piles) {
                total += Math.ceil(pile / k)
            }

            if (total > h) {
                l = k + 1
            } else {
                result = k
                r = k - 1
            }
        }

        return result
    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-5.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        charMap = {}
        result = 0
        l = 0
        currMax = 0

        for r in range(len(s)):
            charMap[s[r]] = 1 + charMap.get(s[r], 0)
            currMax = max(currMax, charMap[s[r]])

            while r - l + 1 - currMax > k:
                charMap[s[l]] -= 1
                l += 1
            
            result = max(currMax, r - l + 1)

        return result 